File: package/umt_go/src/validate/union.go
package validate

// Union creates a union validator that passes when any member validator passes
// (logical OR), mirroring the TypeScript `union` validator
// (Validate/object/union.ts) and the Rust `umt_union` function
// (validate/object/union.rs).
//
// Each member validator returns a ValidateCoreReturnType[E]; only its Validate
// and Message fields are consulted. The returned validator runs the members in
// order and returns a successful result with an empty message as soon as one
// member accepts the value. When no member accepts it, the result is a failure
// carrying the last member's message. In both cases the original value flows back
// through the Type field.
//
// As in the Rust port, Go's type system requires every member to validate the
// same input type T (the TypeScript original accepts heterogeneous validators
// such as `union(string(), number())` over a single dynamic value); E is the
// member result's type-tag parameter and is independent of T.
func Union[T any, E any](validators ...func(value T) ValidateCoreReturnType[E]) func(value T) ValidateCoreReturnType[T] {
	return func(value T) ValidateCoreReturnType[T] {
		lastMessage := ""
		for _, validator := range validators {
			result := validator(value)
			if result.Validate {
				return ValidateCoreReturnType[T]{
					Validate: true,
					Message:  "",
					Type:     value,
				}
			}
			lastMessage = result.Message
		}
		return ValidateCoreReturnType[T]{
			Validate: false,
			Message:  lastMessage,
			Type:     value,
		}
	}
}

File: package/umt_rust/src/validate/object/union.rs
//! union validation utility
//!
//! Mirrors the TypeScript `union` validator, which composes several validators
//! into a single one that passes when *any* of them passes (logical OR). The
//! produced validator runs each member validator in order, returning success
//! with an empty message as soon as one accepts the value, and otherwise
//! propagating the *last* member's failure message.

use super::super::core::ValidateCoreReturnType;

/// Creates a union validator that passes if any of the given validators passes.
///
/// This mirrors the TypeScript `union`. Each member validator returns a
/// [`ValidateCoreReturnType`] (the same `{ validate, message, type }` shape used
/// by every validator in this crate); only its `validate` and `message` fields
/// are consulted. The returned closure runs the members in order: it returns a
/// successful result with an empty message as soon as a member accepts the
/// value, and if none accept it returns a failing result carrying the last
/// member's message. In both cases the original value is returned in
/// `type_value`.
///
/// Unlike the TypeScript original, which accepts heterogeneous validators over a
/// single dynamic value (e.g. `union(string(), number())`), Rust's type system
/// requires every member to validate the same input type `T`, so the members
/// share the closure's parameter type here.
///
/// # Arguments
/// * `validators` - Member validators, each returning a [`ValidateCoreReturnType`]
///
/// # Returns
/// A closure that validates a `T` and returns a `ValidateCoreReturnType<T>`
///
/// # Examples
/// ```
/// use umt_rust::validate::object::umt_union;
/// use umt_rust::validate::core::ValidateCoreReturnType;
///
/// let validator = umt_union(vec![
///     Box::new(|value: &i32| ValidateCoreReturnType {
///         validate: *value < 0,
///         message: "not negative".to_string(),
///         type_value: *value,
///     }) as Box<dyn Fn(&i32) -> ValidateCoreReturnType<i32>>,
///     Box::new(|value: &i32| ValidateCoreReturnType {
///         validate: *value > 10,
///         message: "not greater than ten".to_string(),
///         type_value: *value,
///     }),
/// ]);
///
/// let ok = validator(&-1);
/// assert!(ok.validate);
/// assert_eq!(ok.message, "");
///
/// let bad = validator(&5);
/// assert!(!bad.validate);
/// assert_eq!(bad.message, "not greater than ten");
/// ```
#[allow(clippy::type_complexity)]
pub fn umt_union<T, E>(
    validators: Vec<Box<dyn Fn(&T) -> ValidateCoreReturnType<E>>>,
) -> Box<dyn Fn(&T) -> ValidateCoreReturnType<T>>
where
    T: Clone + 'static,
    E: 'static,
{
    Box::new(move |value: &T| {
        let mut last_message = String::new();
        for validator in &validators {
            let result = validator(value);
            if result.validate {
                return ValidateCoreReturnType {
                    validate: true,
                    message: String::new(),
                    type_value: value.clone(),
                };
            }
            last_message = result.message;
        }

        ValidateCoreReturnType {
            validate: false,
            message: last_message,
            type_value: value.clone(),
        }
    })
}
